client/boot/common.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+import page from 'page';
+import qs from 'querystring';
+import ReactClass from 'react/lib/ReactClass';
+import i18n, { setLocale } from 'i18n-calypso';
+import { some, startsWith } from 'lodash';
+import url from 'url';
+
+/**
+ * Internal dependencies
+ */
+import accessibleFocus from 'lib/accessible-focus';
+import { bindState as bindWpLocaleState } from 'lib/wp/localization';
+import config from 'config';
+import { receiveUser } from 'state/users/actions';
+import {
+	setCurrentUserId,
+	setCurrentUserFlags
+} from 'state/current-user/actions';
+import { setRoute as setRouteAction } from 'state/ui/actions';
+import switchLocale from 'lib/i18n-utils/switch-locale';
+import touchDetect from 'lib/touch-detect';
+
+const debug = debugFactory( 'calypso' );
+
+const switchUserLocale = currentUser => {
+	const localeSlug = currentUser.get().localeSlug;
+	if ( localeSlug ) {
+		switchLocale( localeSlug );
+	}
+};
+
+const setupContextMiddleware = reduxStore => {
+	page( '*', ( context, next ) => {
+		const parsed = url.parse( location.href, true );
+
+		context.store = reduxStore;
+
+		// Break routing and do full
+		//
+		// load for logout link in /me
+		if ( context.pathname === '/wp-login.php' ) {
+			window.location.href = context.path;
+			return;
+		}
+
+		// set `context.query`
+		const querystringStart = context.canonicalPath.indexOf( '?' );
+
+		if ( querystringStart !== -1 ) {
+			context.query = qs.parse( context.canonicalPath.substring( querystringStart + 1 ) );
+		} else {
+			context.query = {};
+		}
+
+		context.prevPath = parsed.path === context.path ? false : parsed.path;
+
+		// set `context.hash` (we have to parse manually)
+		if ( parsed.hash && parsed.hash.length > 1 ) {
+			try {
+				context.hash = qs.parse( parsed.hash.substring( 1 ) );
+			} catch ( e ) {
+				debug( 'failed to query-string parse `location.hash`', e );
+				context.hash = {};
+			}
+		} else {
+			context.hash = {};
+		}
+		next();
+	} );
+};
+
+const loggedOutMiddleware = currentUser => {
+	if ( currentUser.get() ) {
+		return;
+	}
+
+	const sections = require( 'sections' );
+	const validSections = sections.get().reduce( ( acc, section ) => {
+		return section.enableLoggedOut ? acc.concat( section.paths ) : acc;
+	}, [] );
+	const isValidSection = sectionPath => some(
+		validSections, validPath => startsWith( sectionPath, validPath )
+	);
+
+	page( '*', ( context, next ) => {
+		if ( isValidSection( context.path ) ) {
+			next();
+		}
+	} );
+};
+
+const oauthTokenMiddleware = () => {
+	if ( config.isEnabled( 'oauth' ) ) {
+		// Forces OAuth users to the /login page if no token is present
+		page( '*', require( 'auth/controller' ).checkToken );
+	}
+};
+
+const clearNoticesMiddleware = () => {
+	page( '*', function( context, next ) {
```

Comment:
I didn't go for it, because all page handlers call next() at some point. We would have to find a way to signal if page handler breaks the flow or not. In other words if it calls next or returns early. Once we find out how to transition I'm happy to update existing code.